account-kit/smart-contracts/src/light-account/accounts/viem-base.ts

```diff
@@ -0,0 +1,344 @@
+  TLightAccountType,
+  TLightAccountVersion,
+  TTransport,
+  TSigner
+>): Promise<
+  ViemLightAccountBase<TSigner, TLightAccountType, TLightAccountVersion>
+> {
+  const encodeUpgradeToAndCall = async ({
+    upgradeToAddress,
+    upgradeToInitData,
+  }: {
+    upgradeToAddress: Address;
+    upgradeToInitData: Hex;
+  }): Promise<Hex> => {
+    const storage = await getStorageAt(client, {
+      address: accountAddress,
+      slot: "0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc",
+    });
+
+    if (storage == null) {
+      throw new Error(
+        "Failed to get storage slot: 0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc",
+      );
+    }
+
+    const implementationAddresses = Object.values(
+      AccountVersionRegistry[type],
+    ).map(
+      (x) =>
+        x.addresses.overrides?.[client.chain.id]?.impl ??
+        x.addresses.default.impl,
+    );
+
+    if (
+      fromHex(storage, "number") !== 0 &&
+      !implementationAddresses.some((x) => x === trim(storage))
+    ) {
+      throw new Error(
+        `could not determine if smart account implementation is ${type} ${String(
+          version,
+        )}`,
+      );
+    }
+
+    return encodeFunctionData({
+      abi,
+      functionName: "upgradeToAndCall",
+      args: [upgradeToAddress, upgradeToInitData],
+    });
+  };
+
+  const get1271Wrapper = (
+    hashedMessage: Hex,
+    version: string,
+  ): TypedDataDefinition => {
+    return {
+      domain: {
+        chainId: client.chain.id,
+        name: type,
+        verifyingContract: accountAddress,
+        version,
+      },
+      types: {
+        LightAccountMessage: [{ name: "message", type: "bytes" }],
+      },
+      message: {
+        message: hashedMessage,
+      },
+      primaryType: "LightAccountMessage",
+    };
+  };
+
+  const prepareSignature = async (
+    message: string | TypedDataDefinition,
+    isTypedData: boolean = false,
+  ): Promise<{ shouldWrap: boolean; data: string | TypedDataDefinition }> => {
+    const messageHash = isTypedData
+      ? hashTypedData(message as TypedDataDefinition)
+      : hashMessage(message as string);
+
+    switch (version as string) {
+      case "v1.0.1":
+        return { shouldWrap: false, data: message };
+      case "v1.0.2":
+        throw new Error(
+          `Version ${String(version)} of LightAccount doesn't support 1271`,
+        );
+      case "v1.1.0":
+        return { shouldWrap: true, data: get1271Wrapper(messageHash, "1") };
+      case "v2.0.0":
+        return { shouldWrap: true, data: get1271Wrapper(messageHash, "2") };
+      default:
+        throw new Error(`Unknown version ${String(version)} of LightAccount`);
+    }
+  };
+
+  const formatSignature = (signature: Hex): Hex => {
+    return version === "v2.0.0"
+      ? concat([SignatureType.EOA, signature])
+      : signature;
+  };
+
+  const entryPointVersion: EntryPointVersion = (
+    AccountVersionRegistry[type][version] as any
```

Comment:
do we not lint against `any`? 👀

Reply:
we do as params, also yea I couldn't figure out this typing so I gave up